Source/WebKit/Shared/JavaScriptEvaluationResult.mm

```diff
@@ -26,13 +26,225 @@
+        RetainPtr result = adoptNS([[NSMutableDictionary alloc] initWithCapacity:map.size()]);
+        for (auto [keyID, valueID] : map) {
+            RetainPtr key = toID(keyID);
+            if (!key)
+                continue;
+            RetainPtr value = toID(valueID);
+            if (!value)
+                continue;
+            [result setObject:value.get() forKey:key.get()];
+        }
+        return result;
+    });
+}
+
+RetainPtr<id> JavaScriptEvaluationResult::toID(JSObjectID root) const
+{
+    if (RetainPtr object = m_instantiatedObjects.get(root))
+        return object;
+
+    auto it = m_map.find(root);
+    if (it == m_map.end())
+        return nil;
+
+    RetainPtr result = toID(it->value);
+    m_instantiatedObjects.set(root, result);
+    return result;
+}
+
 RetainPtr<id> JavaScriptEvaluationResult::toID() const
 {
-    return API::SerializedScriptValue::deserialize(legacySerializedScriptValue()->internalRepresentation());
+    RetainPtr result = toID(m_root);
+    m_instantiatedObjects.clear();
+    return result;
+}
+
+auto JavaScriptEvaluationResult::toVariant(id object) -> Variant
+{
+    if (!object)
+        return NullType::NullPointer;
+
+    if ([object isKindOfClass:NSNull.class])
+        return NullType::NSNull;
+
+    if ([object isKindOfClass:NSNumber.class]) {
+        if ([object isEqual:@YES])
+            return true;
+        if ([object isEqual:@NO])
+            return false;
+        return CoreIPCNumber((NSNumber *)object);
+    }
+
+    if ([object isKindOfClass:NSString.class])
+        return String((NSString *)object);
+
+    if ([object isKindOfClass:NSDate.class])
+        return Seconds([(NSDate *)object timeIntervalSince1970]);
+
+    if ([object isKindOfClass:NSArray.class]) {
+        Vector<JSObjectID> vector;
+        for (id element : (NSArray *)object)
+            vector.append(addObjectToMap(element));
+        return { WTFMove(vector) };
+    }
+
+    if ([object isKindOfClass:NSDictionary.class]) {
+        HashMap<JSObjectID, JSObjectID> map;
+        [(NSDictionary *)object enumerateKeysAndObjectsUsingBlock:[&](id key, id value, BOOL *) {
+            map.add(addObjectToMap(key), addObjectToMap(value));
+        }];
+        return { WTFMove(map) };
+    }
+
+    // This object has been null checked and came from JSC::valueToObject which only supports these types.
+    ASSERT_NOT_REACHED();
+    return NullType::NullPointer;
+}
+
+JSObjectID JavaScriptEvaluationResult::addObjectToMap(id object)
+{
+    if (!object) {
+        if (!m_nullObjectID) {
+            m_nullObjectID = JSObjectID::generate();
+            m_map.add(*m_nullObjectID, Variant { NullType::NullPointer });
+        }
+        return *m_nullObjectID;
+    }
+
+    auto it = m_objectsInMap.find(object);
+    if (it != m_objectsInMap.end())
+        return it->value;
+
+    auto identifier = JSObjectID::generate();
+    m_map.add(identifier, toVariant(object));
+    m_objectsInMap.set(object, identifier);
```

Comment:
I think we need to update `m_objectsInMap` before `m_map`, otherwise there could be infinite loop for object refers to itself. E.g. for `array` object in `var array = []; array.push(array);`:
1. `addObjectToMap` creates ID=1 for it, since it's not in `m_objectsInMap`
2. `toVariant` converts it by invoking `addObjectToMap` on the first element (so it's the same object)
3. `addObjectToMap` creates ID=2 for it, since it's not in `m_objectsInMap`...

We might want to add a test for this if there isn't one.